keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,338 @@
+---
+title: Coordinated dynamic storage version 
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-08-02
+last-updated: 2019-08-02
+status: provisional
+---
+
+# Coordinated dynamic storage version
+
+## Table of Contents
+
+* [Overview](#overview)
+* [API changes](#api-changes)
+   * [Storage Version API](#storage-version-api)
+* [Changes to API servers](#changes-to-api-servers)
+   * [Curating a list of participating API servers in HA master](#curating-a-list-of-participating-api-servers-in-ha-master)
+   * [Voting for Storage Version](#voting-for-storage-version)
+   * [Voting in action](#voting-in-action)
+   * [The selected storage version is supported by all API servers during upgrades/downgrades](#the-selected-storage-version-is-supported-by-all-api-servers-during-upgradesdowngrades)
+   * [Garbage collection of the StorageVersion objects of removed resources](#garbage-collection-of-the-storageversion-objects-of-removed-resources)
+   * [Using the selected storage version when serializing data for etcd](#using-the-selected-storage-version-when-serializing-data-for-etcd)
+   * [Updating the discovery document](#updating-the-discovery-document)
+   * [CRDs](#crds)
+   * [Aggregated API servers](#aggregated-api-servers)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Risks](#risks)
+* [Graduation Plan](#graduation-plan)
+* [FAQ](#faq)
+* [Future work: solving other coordination problems](#future-work-solving-other-coordination-problems)
+* [Alternatives](#alternatives)
+   * [Exposing if API servers in an HA setup are using the same storage version](#exposing-if-api-servers-in-an-ha-setup-are-using-the-same-storage-version)
+   * [Letting the storage migrator detect if API server instances are in agreement](#letting-the-storage-migrator-detect-if-api-server-instances-are-in-agreement)
+* [References](#references)
+
+## Overview
+
+During the rolling update of an HA master, API server instances *i)* use
+different storage versions for a built-in API resource, and *ii)* show different
+storageVersionHash in the discovery documents. These facts make the storage
```

Comment:
elaborate here.  I suspect it's because etcd may contain records written in different storage versions.